{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b3320f00",
   "metadata": {},
   "source": [
    "# IGVF CRISPR Jamboree 2024: Perturb-seq Inference (Python)\n",
    "\n",
    "Authors: Gene Katsevich and Logan Blaine\n",
    "\n",
    "Date: February 21, 2024\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f33e01c0",
   "metadata": {},
   "source": [
    "# Overview\n",
    "\n",
    "## Perturb-seq inference\n",
    "\n",
    "The goal of perturb-seq inference is to quantify the extent to which the perturbation of a given genomic element impacts the expression of a given gene. We allow a range of statistical interpretations of this task. In a frequentist framework, this task can be viewed as testing the null hypothesis that the perturbation of the genomic element has no effect on the gene's expression, or as estimating the effect size of the perturbation on the gene's expression. In a Bayesian framework, this task can be viewed as estimating the posterior probability of the presence of a non-zero effect, or as a posterior mean of the effect size. \n",
    "\n",
    "## Jamboree goals\n",
    "\n",
    "The goal of the perturb-seq inference portion of the Jamboree is to implement a number of perturb-seq inference methodologies using common input and output formats. Following the Jamboree, these implementations will be added as modules to a Nextflow pipeline, which will then be used to benchmark their statistical and computational performance. This benchmarking effort will suggest best practices for perturb-seq inference, and will be used to inform the development of the IGVF perturb-seq analysis pipeline.\n",
    "\n",
    "## Data format overview\n",
    "\n",
    "The primary input to a perturb-seq inference module is a `MuData` object, which contains both the perturb-seq data and a set of element-gene pairs for which the inference is to be performed. The output of each method should be the same `MuData` object, except with an additional table containing one or more measures of association for each element-gene pair. The `MuData` format is an HDF5-based language-agnostic format compatible with import into both R and Python. Each `MuData` object will contain a minimal set of fields required for inference, and potentially one or more optional fields that provide additional information. For the purposes of this Jamboree, we have provided `MuData` objects for subsets of the Gasperini et al (2019) and Papalexi et al (2021) datasets. For each dataset, we have provided a minimal `MuData` object that contains just the required fields, as well as a more fleshed out object that contains additional optional fields.\n",
    "\n",
    "## Requested function API, documentation, and demonstration\n",
    "\n",
    "Please write a function in your language of choice with the following arguments: \n",
    "\n",
    "- The first argument should be `mudata_input_fp`, a filepath to a `MuData` object.\n",
    "- The second argument should be `mudata_output_fp`, a filepath to an output `MuData` object.\n",
    "- There may be one or more additional arguments specific to your method. \n",
    "\n",
    "The function should read the `MuData` object from `mudata_input_fp`, perform the inference, and write the resulting `MuData` object to `mudata_output_fp` (in Python, via `mudata.read_h5mu()`). The function should include documentation of any additional arguments used. Furthermore, please include a demonstration of the use of your function on at least one of the sample datasets provided, and a brief discussion of the results.\n",
    "\n",
    "# `MuData` format\n",
    "\n",
    "Let us walk through the input and output format specifications, from the perspective of Python, using a subset of the Gasperini et al (2019) dataset as an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "12c3505d",
   "metadata": {
    "warning": false
   },
   "outputs": [],
   "source": [
    "import mudata as md\n",
    "import pandas as pd\n",
    "\n",
    "data_dir = \"/Users/ljb80/Projects/igvf-jamboree/sceptre-igvf/data/inference\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "768a5f2e",
   "metadata": {},
   "source": [
    "## Required input fields\n",
    "\n",
    "We start with an example of the minimal `MuData` object required for perturb-seq inference."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "af24cd51",
   "metadata": {
    "lines_to_next_cell": 0,
    "warning": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 9704 × 167\n",
       "  uns:\t&#x27;pairs_to_test&#x27;\n",
       "  2 modalities\n",
       "    gene:\t9704 x 112\n",
       "    guide:\t9704 x 55\n",
       "      var:\t&#x27;targeting&#x27;, &#x27;intended_target_name&#x27;\n",
       "      uns:\t&#x27;capture_method&#x27;, &#x27;moi&#x27;\n",
       "      layers:\t&#x27;guide_assignment&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 9704 × 167\n",
       "  uns:\t'pairs_to_test'\n",
       "  2 modalities\n",
       "    gene:\t9704 x 112\n",
       "    guide:\t9704 x 55\n",
       "      var:\t'targeting', 'intended_target_name'\n",
       "      uns:\t'capture_method', 'moi'\n",
       "      layers:\t'guide_assignment'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mudata_input_fp = f\"{data_dir}/gasperini_inference_input_minimal.h5mu\"\n",
    "input_minimal = md.read_h5mu(mudata_input_fp)\n",
    "input_minimal"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbe980bf",
   "metadata": {},
   "source": [
    "The minimal `MuData` object for perturb-seq inference contains two modalities: `gene` and `guide`. \n",
    "\n",
    "### `gene` modality\n",
    "\n",
    "The `gene` modality just needs to have a `.X` matrix containing the RNA UMI counts. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7969ebc2",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000008853</th>\n",
       "      <th>ENSG00000104679</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCTTGAATCGAATGCT-1_1B_1_SI-GA-F2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGCTTGATCGAGAGCA-1_1A_2_SI-GA-E3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  ENSG00000008853  ENSG00000104679\n",
       "GCTTGAATCGAATGCT-1_1B_1_SI-GA-F2              0.0              2.0\n",
       "AGCTTGATCGAGAGCA-1_1A_2_SI-GA-E3              0.0              0.0"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_minimal[\"gene\"].X\n",
    "input_minimal[\"gene\"][:2, :2].to_df()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6a1c538",
   "metadata": {},
   "source": [
    "### `guide` modality\n",
    "\n",
    "The `guide` modality needs to have a `.X` matrix containing the gRNA UMI counts, as well as a `.layers['guide_assignment']` matrix containing the binary gRNA assignments. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4ccf971b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<9704x55 sparse matrix of type '<class 'numpy.float64'>'\n",
       "\twith 10563 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_minimal[\"guide\"].X\n",
    "input_minimal[\"guide\"].layers[\"guide_assignment\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cbde1ed",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "We can view a couple rows and columns of each:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2761e35a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GCCCTGCTACCCACTTACAG</th>\n",
       "      <th>ATGTAGAAGGAGACACCGGG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCTTGAATCGAATGCT-1_1B_1_SI-GA-F2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGTGAAGCACCAGGCT-1_1A_6_SI-GA-E7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  GCCCTGCTACCCACTTACAG  ATGTAGAAGGAGACACCGGG\n",
       "GCTTGAATCGAATGCT-1_1B_1_SI-GA-F2                   1.0                   0.0\n",
       "GGTGAAGCACCAGGCT-1_1A_6_SI-GA-E7                   0.0                   1.0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_ids = [\"GCTTGAATCGAATGCT-1_1B_1_SI-GA-F2\", \"GGTGAAGCACCAGGCT-1_1A_6_SI-GA-E7\"]\n",
    "grna_ids = [\"GCCCTGCTACCCACTTACAG\", \"ATGTAGAAGGAGACACCGGG\"]\n",
    "pd.DataFrame(\n",
    "    input_minimal[\"guide\"][cell_ids, grna_ids].X.toarray(),\n",
    "    index=cell_ids,\n",
    "    columns=grna_ids,\n",
    ")\n",
    "\n",
    "pd.DataFrame(\n",
    "    input_minimal[\"guide\"][cell_ids, grna_ids].layers[\"guide_assignment\"].toarray(),\n",
    "    index=cell_ids,\n",
    "    columns=grna_ids,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45f189fc",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "In addition to the guide UMI counts and assignments, the `guide` modality must contain certain metadata information. This includes a `.var` data frame containing at least the binary variable `targeting` (`TRUE` if the guide targets a genomic element of interest or `FALSE` if it is safe- or non-targeting) and the string `intended_target_name` (the name of the genomic element targeted by the guide). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "29d9f627",
   "metadata": {
    "lines_to_next_cell": 0,
    "warning": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>targeting</th>\n",
       "      <th>intended_target_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ATGTAGAAGGAGACACCGGG</th>\n",
       "      <td>TRUE</td>\n",
       "      <td>ENSG00000012660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCGCAGAGGCGGATGTAGAG</th>\n",
       "      <td>TRUE</td>\n",
       "      <td>ENSG00000012660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACACCCTCATTAGAACCCAG</th>\n",
       "      <td>TRUE</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTAAGAGCCTCGGTTCCCCT</th>\n",
       "      <td>TRUE</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GACCTCCTGTGATCAGGTGG</th>\n",
       "      <td>FALSE</td>\n",
       "      <td>non-targeting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATTGGTATCCGTATAAGCAG</th>\n",
       "      <td>FALSE</td>\n",
       "      <td>non-targeting</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     targeting intended_target_name\n",
       "ATGTAGAAGGAGACACCGGG      TRUE      ENSG00000012660\n",
       "GCGCAGAGGCGGATGTAGAG      TRUE      ENSG00000012660\n",
       "ACACCCTCATTAGAACCCAG      TRUE      candidate_enh_1\n",
       "TTAAGAGCCTCGGTTCCCCT      TRUE      candidate_enh_1\n",
       "GACCTCCTGTGATCAGGTGG     FALSE        non-targeting\n",
       "ATTGGTATCCGTATAAGCAG     FALSE        non-targeting"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_minimal[\"guide\"].var.iloc[[0, 1, 20, 21, 30, 31]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d96d21f0",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "Note that the `targeting` column is a string rather than a Boolean due to type compatibility issues involving R, Python, and HDF5. It can be cast to a Boolean if desired. \n",
    "\n",
    "Finally, the guide modality must contain `uns` fields called `moi` (`low` or `high`) and `capture_method` (`\"CROP-seq\"` or `\"direct capture\"`):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "16dcb9b7",
   "metadata": {
    "warning": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'high'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_minimal[\"guide\"].uns[\"capture_method\"][0]\n",
    "input_minimal[\"guide\"].uns[\"moi\"][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6de94e5",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Global `.uns`\n",
    "\n",
    "The input `MuData` object is also required to have a global `.uns` field named `pairs_to_test`, which is a data frame containing the pairs of elements (specified via `intended_target_name`) and genes (specified via `gene_id`) for which the inference is to be performed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4b6e4aa2",
   "metadata": {
    "warning": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>intended_target_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000187109</td>\n",
       "      <td>ENSG00000187109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000114850</td>\n",
       "      <td>ENSG00000114850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000134851</td>\n",
       "      <td>ENSG00000134851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000163866</td>\n",
       "      <td>ENSG00000163866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000181610</td>\n",
       "      <td>ENSG00000181610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>ENSG00000106789</td>\n",
       "      <td>candidate_enh_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>ENSG00000125482</td>\n",
       "      <td>candidate_enh_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>ENSG00000095380</td>\n",
       "      <td>candidate_enh_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>ENSG00000158941</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>ENSG00000167123</td>\n",
       "      <td>candidate_enh_3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_id intended_target_name\n",
       "0    ENSG00000187109      ENSG00000187109\n",
       "1    ENSG00000114850      ENSG00000114850\n",
       "2    ENSG00000134851      ENSG00000134851\n",
       "3    ENSG00000163866      ENSG00000163866\n",
       "4    ENSG00000181610      ENSG00000181610\n",
       "..               ...                  ...\n",
       "105  ENSG00000106789      candidate_enh_2\n",
       "106  ENSG00000125482      candidate_enh_3\n",
       "107  ENSG00000095380      candidate_enh_2\n",
       "108  ENSG00000158941      candidate_enh_1\n",
       "109  ENSG00000167123      candidate_enh_3\n",
       "\n",
       "[110 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(input_minimal.uns[\"pairs_to_test\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b78f06cc",
   "metadata": {},
   "source": [
    "## Optional input fields\n",
    "\n",
    "Next we consider optional fields that can be included in the input `MuData` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d38ad0fa",
   "metadata": {
    "lines_to_next_cell": 0,
    "warning": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 9704 × 167\n",
       "  obs:\t&#x27;prep_batch&#x27;, &#x27;within_batch_chip&#x27;, &#x27;within_chip_lane&#x27;\n",
       "  uns:\t&#x27;pairs_to_test&#x27;\n",
       "  2 modalities\n",
       "    gene:\t9704 x 112\n",
       "      obs:\t&#x27;num_expressed_genes&#x27;, &#x27;total_gene_umis&#x27;\n",
       "      var:\t&#x27;symbol&#x27;, &#x27;gene_chr&#x27;, &#x27;gene_start&#x27;, &#x27;gene_end&#x27;\n",
       "    guide:\t9704 x 55\n",
       "      obs:\t&#x27;num_expressed_guides&#x27;, &#x27;total_guide_umis&#x27;\n",
       "      var:\t&#x27;targeting&#x27;, &#x27;intended_target_name&#x27;, &#x27;intended_target_chr&#x27;, &#x27;intended_target_start&#x27;, &#x27;intended_target_end&#x27;\n",
       "      uns:\t&#x27;capture_method&#x27;, &#x27;moi&#x27;\n",
       "      layers:\t&#x27;guide_assignment&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 9704 × 167\n",
       "  obs:\t'prep_batch', 'within_batch_chip', 'within_chip_lane'\n",
       "  uns:\t'pairs_to_test'\n",
       "  2 modalities\n",
       "    gene:\t9704 x 112\n",
       "      obs:\t'num_expressed_genes', 'total_gene_umis'\n",
       "      var:\t'symbol', 'gene_chr', 'gene_start', 'gene_end'\n",
       "    guide:\t9704 x 55\n",
       "      obs:\t'num_expressed_guides', 'total_guide_umis'\n",
       "      var:\t'targeting', 'intended_target_name', 'intended_target_chr', 'intended_target_start', 'intended_target_end'\n",
       "      uns:\t'capture_method', 'moi'\n",
       "      layers:\t'guide_assignment'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mudata_input_fp = f\"{data_dir}/gasperini_inference_input.h5mu\"\n",
    "input_optional = md.read_h5mu(mudata_input_fp)\n",
    "input_optional"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cae75433",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### `gene` modality\n",
    "\n",
    "The `MuData` object may include cellwise covariates for the `gene` modality in `.mod['gene].obs`, such as number of genes with nonzero UMI counts (`num_expressed_genes`) and total RNA UMIs (`total_gene_umis`):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a08eec29",
   "metadata": {
    "warning": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_expressed_genes</th>\n",
       "      <th>total_gene_umis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCTTGAATCGAATGCT-1_1B_1_SI-GA-F2</th>\n",
       "      <td>41</td>\n",
       "      <td>280.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGCTTGATCGAGAGCA-1_1A_2_SI-GA-E3</th>\n",
       "      <td>35</td>\n",
       "      <td>192.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCCAATCTCCTCAATT-1_1B_1_SI-GA-F2</th>\n",
       "      <td>41</td>\n",
       "      <td>781.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCGGTACACTGTCGG-1_1A_2_SI-GA-E3</th>\n",
       "      <td>37</td>\n",
       "      <td>189.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGACGTCTCATGTCTT-1_1B_8_SI-GA-F9</th>\n",
       "      <td>32</td>\n",
       "      <td>262.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCTATCTCTATCGCC-1_2A_4_SI-GA-G5</th>\n",
       "      <td>23</td>\n",
       "      <td>203.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCACAAGCAGCCTTGG-1_2A_6_SI-GA-G7</th>\n",
       "      <td>30</td>\n",
       "      <td>173.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGCAGGTGAAGGCT-1_2B_6_SI-GA-H7</th>\n",
       "      <td>37</td>\n",
       "      <td>428.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGATTACCATGTTGAC-1_2A_4_SI-GA-G5</th>\n",
       "      <td>47</td>\n",
       "      <td>658.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCTTCTCGGATGTT-1_2A_1_SI-GA-G2</th>\n",
       "      <td>23</td>\n",
       "      <td>166.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9704 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  num_expressed_genes  total_gene_umis\n",
       "GCTTGAATCGAATGCT-1_1B_1_SI-GA-F2                   41            280.0\n",
       "AGCTTGATCGAGAGCA-1_1A_2_SI-GA-E3                   35            192.0\n",
       "CCCAATCTCCTCAATT-1_1B_1_SI-GA-F2                   41            781.0\n",
       "CGCGGTACACTGTCGG-1_1A_2_SI-GA-E3                   37            189.0\n",
       "GGACGTCTCATGTCTT-1_1B_8_SI-GA-F9                   32            262.0\n",
       "...                                               ...              ...\n",
       "CGCTATCTCTATCGCC-1_2A_4_SI-GA-G5                   23            203.0\n",
       "TCACAAGCAGCCTTGG-1_2A_6_SI-GA-G7                   30            173.0\n",
       "GCTGCAGGTGAAGGCT-1_2B_6_SI-GA-H7                   37            428.0\n",
       "GGATTACCATGTTGAC-1_2A_4_SI-GA-G5                   47            658.0\n",
       "GTGCTTCTCGGATGTT-1_2A_1_SI-GA-G2                   23            166.0\n",
       "\n",
       "[9704 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_optional[\"gene\"].obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63cc241f",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "The `MuData` object may include per-gene metadata in `.mod['gene'].var`, such as the HGNC gene symbol (`symbol`), the gene chromosome (`chr`), start (`gene_start`), and end (`gene_end`) coordinates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "87a38bd6",
   "metadata": {
    "warning": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>gene_chr</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000008853</th>\n",
       "      <td>RHOBTB2</td>\n",
       "      <td>chr8</td>\n",
       "      <td>22844930.0</td>\n",
       "      <td>22844931.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000104679</th>\n",
       "      <td>R3HCC1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>23145421.0</td>\n",
       "      <td>23145422.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000104689</th>\n",
       "      <td>TNFRSF10A</td>\n",
       "      <td>chr8</td>\n",
       "      <td>23082573.0</td>\n",
       "      <td>23082574.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000120889</th>\n",
       "      <td>TNFRSF10B</td>\n",
       "      <td>chr8</td>\n",
       "      <td>22926533.0</td>\n",
       "      <td>22926534.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000120896</th>\n",
       "      <td>SORBS3</td>\n",
       "      <td>chr8</td>\n",
       "      <td>22409208.0</td>\n",
       "      <td>22409209.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000114850</th>\n",
       "      <td>SSR3</td>\n",
       "      <td>chr3</td>\n",
       "      <td>156271913.0</td>\n",
       "      <td>156271914.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000072274</th>\n",
       "      <td>TFRC</td>\n",
       "      <td>chr3</td>\n",
       "      <td>195808960.0</td>\n",
       "      <td>195808961.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000134851</th>\n",
       "      <td>TMEM165</td>\n",
       "      <td>chr4</td>\n",
       "      <td>56262124.0</td>\n",
       "      <td>56262125.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198899</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228253</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    symbol gene_chr   gene_start     gene_end\n",
       "ENSG00000008853    RHOBTB2     chr8   22844930.0   22844931.0\n",
       "ENSG00000104679     R3HCC1     chr8   23145421.0   23145422.0\n",
       "ENSG00000104689  TNFRSF10A     chr8   23082573.0   23082574.0\n",
       "ENSG00000120889  TNFRSF10B     chr8   22926533.0   22926534.0\n",
       "ENSG00000120896     SORBS3     chr8   22409208.0   22409209.0\n",
       "...                    ...      ...          ...          ...\n",
       "ENSG00000114850       SSR3     chr3  156271913.0  156271914.0\n",
       "ENSG00000072274       TFRC     chr3  195808960.0  195808961.0\n",
       "ENSG00000134851    TMEM165     chr4   56262124.0   56262125.0\n",
       "ENSG00000198899                              NaN          NaN\n",
       "ENSG00000228253                              NaN          NaN\n",
       "\n",
       "[112 rows x 4 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_optional[\"gene\"].var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be2e09c1",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### `guide` modality\n",
    "\n",
    "The `MuData` object may include cellwise covariates for the `guide` modality in `.mod['guide'].obs`, such as number of guides with nonzero UMI counts (`num_expressed_guides`) and total guide UMIs (`total_guide_umis`):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e1f7a2d6",
   "metadata": {
    "warning": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_expressed_guides</th>\n",
       "      <th>total_guide_umis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCTTGAATCGAATGCT-1_1B_1_SI-GA-F2</th>\n",
       "      <td>1</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGCTTGATCGAGAGCA-1_1A_2_SI-GA-E3</th>\n",
       "      <td>1</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCCAATCTCCTCAATT-1_1B_1_SI-GA-F2</th>\n",
       "      <td>1</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCGGTACACTGTCGG-1_1A_2_SI-GA-E3</th>\n",
       "      <td>1</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGACGTCTCATGTCTT-1_1B_8_SI-GA-F9</th>\n",
       "      <td>1</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCTATCTCTATCGCC-1_2A_4_SI-GA-G5</th>\n",
       "      <td>1</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCACAAGCAGCCTTGG-1_2A_6_SI-GA-G7</th>\n",
       "      <td>1</td>\n",
       "      <td>39.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGCAGGTGAAGGCT-1_2B_6_SI-GA-H7</th>\n",
       "      <td>1</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGATTACCATGTTGAC-1_2A_4_SI-GA-G5</th>\n",
       "      <td>1</td>\n",
       "      <td>73.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCTTCTCGGATGTT-1_2A_1_SI-GA-G2</th>\n",
       "      <td>1</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9704 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  num_expressed_guides  total_guide_umis\n",
       "GCTTGAATCGAATGCT-1_1B_1_SI-GA-F2                     1               9.0\n",
       "AGCTTGATCGAGAGCA-1_1A_2_SI-GA-E3                     1              18.0\n",
       "CCCAATCTCCTCAATT-1_1B_1_SI-GA-F2                     1              24.0\n",
       "CGCGGTACACTGTCGG-1_1A_2_SI-GA-E3                     1              26.0\n",
       "GGACGTCTCATGTCTT-1_1B_8_SI-GA-F9                     1              12.0\n",
       "...                                                ...               ...\n",
       "CGCTATCTCTATCGCC-1_2A_4_SI-GA-G5                     1               5.0\n",
       "TCACAAGCAGCCTTGG-1_2A_6_SI-GA-G7                     1              39.0\n",
       "GCTGCAGGTGAAGGCT-1_2B_6_SI-GA-H7                     1              21.0\n",
       "GGATTACCATGTTGAC-1_2A_4_SI-GA-G5                     1              73.0\n",
       "GTGCTTCTCGGATGTT-1_2A_1_SI-GA-G2                     1              12.0\n",
       "\n",
       "[9704 rows x 2 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_optional[\"guide\"].obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89de6675",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "The `MuData` object may include per-guide metadata in `.mod['guide'].var` in addition to the required `targeting` and `intended_target_name` fields, such as the chromosome (`intended_target_chr`), start (`intended_target_start`), and end (`intended_target_end`) of the targeted element:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3028666d",
   "metadata": {
    "warning": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>targeting</th>\n",
       "      <th>intended_target_name</th>\n",
       "      <th>intended_target_chr</th>\n",
       "      <th>intended_target_start</th>\n",
       "      <th>intended_target_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ATGTAGAAGGAGACACCGGG</th>\n",
       "      <td>TRUE</td>\n",
       "      <td>ENSG00000012660</td>\n",
       "      <td>chr6</td>\n",
       "      <td>53213723.0</td>\n",
       "      <td>53213738.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCGCAGAGGCGGATGTAGAG</th>\n",
       "      <td>TRUE</td>\n",
       "      <td>ENSG00000012660</td>\n",
       "      <td>chr6</td>\n",
       "      <td>53213738.0</td>\n",
       "      <td>53213754.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACACCCTCATTAGAACCCAG</th>\n",
       "      <td>TRUE</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>23366136.0</td>\n",
       "      <td>23366564.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTAAGAGCCTCGGTTCCCCT</th>\n",
       "      <td>TRUE</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>23366564.0</td>\n",
       "      <td>23366992.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GACCTCCTGTGATCAGGTGG</th>\n",
       "      <td>FALSE</td>\n",
       "      <td>non-targeting</td>\n",
       "      <td></td>\n",
       "      <td>-9.0</td>\n",
       "      <td>-9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATTGGTATCCGTATAAGCAG</th>\n",
       "      <td>FALSE</td>\n",
       "      <td>non-targeting</td>\n",
       "      <td></td>\n",
       "      <td>-9.0</td>\n",
       "      <td>-9.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     targeting intended_target_name intended_target_chr  \\\n",
       "ATGTAGAAGGAGACACCGGG      TRUE      ENSG00000012660                chr6   \n",
       "GCGCAGAGGCGGATGTAGAG      TRUE      ENSG00000012660                chr6   \n",
       "ACACCCTCATTAGAACCCAG      TRUE      candidate_enh_1                chr8   \n",
       "TTAAGAGCCTCGGTTCCCCT      TRUE      candidate_enh_1                chr8   \n",
       "GACCTCCTGTGATCAGGTGG     FALSE        non-targeting                       \n",
       "ATTGGTATCCGTATAAGCAG     FALSE        non-targeting                       \n",
       "\n",
       "                      intended_target_start  intended_target_end  \n",
       "ATGTAGAAGGAGACACCGGG             53213723.0           53213738.0  \n",
       "GCGCAGAGGCGGATGTAGAG             53213738.0           53213754.0  \n",
       "ACACCCTCATTAGAACCCAG             23366136.0           23366564.0  \n",
       "TTAAGAGCCTCGGTTCCCCT             23366564.0           23366992.0  \n",
       "GACCTCCTGTGATCAGGTGG                   -9.0                 -9.0  \n",
       "ATTGGTATCCGTATAAGCAG                   -9.0                 -9.0  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_optional[\"guide\"].var.iloc[[0, 1, 20, 21, 30, 31]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68ae01dc",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Global `.obs`\n",
    "\n",
    "Optionally, the `MuData` input object can contain a global `obs` field containing cell-level information that is not specific to modality, such as batch information. Here is what it looks like for the Gasperini data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cf8f5af2",
   "metadata": {
    "warning": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prep_batch</th>\n",
       "      <th>within_batch_chip</th>\n",
       "      <th>within_chip_lane</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCTTGAATCGAATGCT-1_1B_1_SI-GA-F2</th>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_B</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGCTTGATCGAGAGCA-1_1A_2_SI-GA-E3</th>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCCAATCTCCTCAATT-1_1B_1_SI-GA-F2</th>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_B</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCGGTACACTGTCGG-1_1A_2_SI-GA-E3</th>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGACGTCTCATGTCTT-1_1B_8_SI-GA-F9</th>\n",
       "      <td>prep_batch_1</td>\n",
       "      <td>within_batch_chip_B</td>\n",
       "      <td>within_chip_lane_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGCTATCTCTATCGCC-1_2A_4_SI-GA-G5</th>\n",
       "      <td>prep_batch_2</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCACAAGCAGCCTTGG-1_2A_6_SI-GA-G7</th>\n",
       "      <td>prep_batch_2</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCTGCAGGTGAAGGCT-1_2B_6_SI-GA-H7</th>\n",
       "      <td>prep_batch_2</td>\n",
       "      <td>within_batch_chip_B</td>\n",
       "      <td>within_chip_lane_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGATTACCATGTTGAC-1_2A_4_SI-GA-G5</th>\n",
       "      <td>prep_batch_2</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGCTTCTCGGATGTT-1_2A_1_SI-GA-G2</th>\n",
       "      <td>prep_batch_2</td>\n",
       "      <td>within_batch_chip_A</td>\n",
       "      <td>within_chip_lane_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9704 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    prep_batch    within_batch_chip  \\\n",
       "GCTTGAATCGAATGCT-1_1B_1_SI-GA-F2  prep_batch_1  within_batch_chip_B   \n",
       "AGCTTGATCGAGAGCA-1_1A_2_SI-GA-E3  prep_batch_1  within_batch_chip_A   \n",
       "CCCAATCTCCTCAATT-1_1B_1_SI-GA-F2  prep_batch_1  within_batch_chip_B   \n",
       "CGCGGTACACTGTCGG-1_1A_2_SI-GA-E3  prep_batch_1  within_batch_chip_A   \n",
       "GGACGTCTCATGTCTT-1_1B_8_SI-GA-F9  prep_batch_1  within_batch_chip_B   \n",
       "...                                        ...                  ...   \n",
       "CGCTATCTCTATCGCC-1_2A_4_SI-GA-G5  prep_batch_2  within_batch_chip_A   \n",
       "TCACAAGCAGCCTTGG-1_2A_6_SI-GA-G7  prep_batch_2  within_batch_chip_A   \n",
       "GCTGCAGGTGAAGGCT-1_2B_6_SI-GA-H7  prep_batch_2  within_batch_chip_B   \n",
       "GGATTACCATGTTGAC-1_2A_4_SI-GA-G5  prep_batch_2  within_batch_chip_A   \n",
       "GTGCTTCTCGGATGTT-1_2A_1_SI-GA-G2  prep_batch_2  within_batch_chip_A   \n",
       "\n",
       "                                    within_chip_lane  \n",
       "GCTTGAATCGAATGCT-1_1B_1_SI-GA-F2  within_chip_lane_1  \n",
       "AGCTTGATCGAGAGCA-1_1A_2_SI-GA-E3  within_chip_lane_2  \n",
       "CCCAATCTCCTCAATT-1_1B_1_SI-GA-F2  within_chip_lane_1  \n",
       "CGCGGTACACTGTCGG-1_1A_2_SI-GA-E3  within_chip_lane_2  \n",
       "GGACGTCTCATGTCTT-1_1B_8_SI-GA-F9  within_chip_lane_8  \n",
       "...                                              ...  \n",
       "CGCTATCTCTATCGCC-1_2A_4_SI-GA-G5  within_chip_lane_4  \n",
       "TCACAAGCAGCCTTGG-1_2A_6_SI-GA-G7  within_chip_lane_6  \n",
       "GCTGCAGGTGAAGGCT-1_2B_6_SI-GA-H7  within_chip_lane_6  \n",
       "GGATTACCATGTTGAC-1_2A_4_SI-GA-G5  within_chip_lane_4  \n",
       "GTGCTTCTCGGATGTT-1_2A_1_SI-GA-G2  within_chip_lane_1  \n",
       "\n",
       "[9704 rows x 3 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_optional.obs[[\"prep_batch\", \"within_batch_chip\", \"within_chip_lane\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42d38746",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "### Pairs to test \n",
    "\n",
    "Optionally, `.uns['pairs_to_test']` can have a third column: `pair_type`: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "09030940",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>intended_target_name</th>\n",
       "      <th>pair_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000187109</td>\n",
       "      <td>ENSG00000187109</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000114850</td>\n",
       "      <td>ENSG00000114850</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000134851</td>\n",
       "      <td>ENSG00000134851</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000163866</td>\n",
       "      <td>ENSG00000163866</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000181610</td>\n",
       "      <td>ENSG00000181610</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>ENSG00000106789</td>\n",
       "      <td>candidate_enh_2</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>ENSG00000125482</td>\n",
       "      <td>candidate_enh_3</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>ENSG00000095380</td>\n",
       "      <td>candidate_enh_2</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>ENSG00000158941</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>ENSG00000167123</td>\n",
       "      <td>candidate_enh_3</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_id intended_target_name         pair_type\n",
       "0    ENSG00000187109      ENSG00000187109  positive_control\n",
       "1    ENSG00000114850      ENSG00000114850  positive_control\n",
       "2    ENSG00000134851      ENSG00000134851  positive_control\n",
       "3    ENSG00000163866      ENSG00000163866  positive_control\n",
       "4    ENSG00000181610      ENSG00000181610  positive_control\n",
       "..               ...                  ...               ...\n",
       "105  ENSG00000106789      candidate_enh_2         discovery\n",
       "106  ENSG00000125482      candidate_enh_3         discovery\n",
       "107  ENSG00000095380      candidate_enh_2         discovery\n",
       "108  ENSG00000158941      candidate_enh_1         discovery\n",
       "109  ENSG00000167123      candidate_enh_3         discovery\n",
       "\n",
       "[110 rows x 3 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(input_optional.uns[\"pairs_to_test\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee5fe0f8",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "This optional column classifies pairs based on whether they are intended to be positive controls (an association is known to exist), negative controls (an association is known not to exist), or discovery pairs (pairs where it is unknown whether an association exists). This information need not be used by the inference module, but it is useful for downstream analysis.\n",
    "\n",
    "## Output fields\n",
    "\n",
    "The output should be the same `MuData` object as the input, with the addition of a `test_results` field to the global `.uns`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5cad2f5f",
   "metadata": {
    "warning": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 9704 × 167\n",
       "  obs:\t&#x27;prep_batch&#x27;, &#x27;within_batch_chip&#x27;, &#x27;within_chip_lane&#x27;\n",
       "  uns:\t&#x27;pairs_to_test&#x27;, &#x27;test_results&#x27;\n",
       "  2 modalities\n",
       "    gene:\t9704 x 112\n",
       "      obs:\t&#x27;num_expressed_genes&#x27;, &#x27;total_gene_umis&#x27;\n",
       "      var:\t&#x27;symbol&#x27;, &#x27;gene_chr&#x27;, &#x27;gene_start&#x27;, &#x27;gene_end&#x27;\n",
       "    guide:\t9704 x 55\n",
       "      obs:\t&#x27;num_expressed_guides&#x27;, &#x27;total_guide_umis&#x27;\n",
       "      var:\t&#x27;targeting&#x27;, &#x27;intended_target_name&#x27;, &#x27;intended_target_chr&#x27;, &#x27;intended_target_start&#x27;, &#x27;intended_target_end&#x27;\n",
       "      uns:\t&#x27;capture_method&#x27;, &#x27;moi&#x27;\n",
       "      layers:\t&#x27;guide_assignment&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 9704 × 167\n",
       "  obs:\t'prep_batch', 'within_batch_chip', 'within_chip_lane'\n",
       "  uns:\t'pairs_to_test', 'test_results'\n",
       "  2 modalities\n",
       "    gene:\t9704 x 112\n",
       "      obs:\t'num_expressed_genes', 'total_gene_umis'\n",
       "      var:\t'symbol', 'gene_chr', 'gene_start', 'gene_end'\n",
       "    guide:\t9704 x 55\n",
       "      obs:\t'num_expressed_guides', 'total_guide_umis'\n",
       "      var:\t'targeting', 'intended_target_name', 'intended_target_chr', 'intended_target_start', 'intended_target_end'\n",
       "      uns:\t'capture_method', 'moi'\n",
       "      layers:\t'guide_assignment'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mudata_output_fp = f\"{data_dir}/gasperini_inference_output.h5mu\"\n",
    "output_optional = md.read_h5mu(mudata_output_fp)\n",
    "output_optional"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "66873b71",
   "metadata": {
    "lines_to_next_cell": 0,
    "warning": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>intended_target_name</th>\n",
       "      <th>log2_fc</th>\n",
       "      <th>p_value</th>\n",
       "      <th>pair_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000187109</td>\n",
       "      <td>ENSG00000187109</td>\n",
       "      <td>-0.774367</td>\n",
       "      <td>3.217223e-85</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000114850</td>\n",
       "      <td>ENSG00000114850</td>\n",
       "      <td>-1.849572</td>\n",
       "      <td>2.414163e-79</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000134851</td>\n",
       "      <td>ENSG00000134851</td>\n",
       "      <td>-0.893860</td>\n",
       "      <td>4.309833e-50</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000163866</td>\n",
       "      <td>ENSG00000163866</td>\n",
       "      <td>-1.223700</td>\n",
       "      <td>4.704066e-49</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000181610</td>\n",
       "      <td>ENSG00000181610</td>\n",
       "      <td>-1.314285</td>\n",
       "      <td>3.766690e-42</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>ENSG00000106789</td>\n",
       "      <td>candidate_enh_2</td>\n",
       "      <td>0.079632</td>\n",
       "      <td>6.660000e-01</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>ENSG00000125482</td>\n",
       "      <td>candidate_enh_3</td>\n",
       "      <td>0.144014</td>\n",
       "      <td>8.900000e-01</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>ENSG00000095380</td>\n",
       "      <td>candidate_enh_2</td>\n",
       "      <td>-0.165492</td>\n",
       "      <td>3.400000e-02</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>ENSG00000158941</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "      <td>0.117617</td>\n",
       "      <td>7.980000e-01</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>ENSG00000167123</td>\n",
       "      <td>candidate_enh_3</td>\n",
       "      <td>-0.482057</td>\n",
       "      <td>8.800000e-02</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_id intended_target_name   log2_fc       p_value  \\\n",
       "0    ENSG00000187109      ENSG00000187109 -0.774367  3.217223e-85   \n",
       "1    ENSG00000114850      ENSG00000114850 -1.849572  2.414163e-79   \n",
       "2    ENSG00000134851      ENSG00000134851 -0.893860  4.309833e-50   \n",
       "3    ENSG00000163866      ENSG00000163866 -1.223700  4.704066e-49   \n",
       "4    ENSG00000181610      ENSG00000181610 -1.314285  3.766690e-42   \n",
       "..               ...                  ...       ...           ...   \n",
       "105  ENSG00000106789      candidate_enh_2  0.079632  6.660000e-01   \n",
       "106  ENSG00000125482      candidate_enh_3  0.144014  8.900000e-01   \n",
       "107  ENSG00000095380      candidate_enh_2 -0.165492  3.400000e-02   \n",
       "108  ENSG00000158941      candidate_enh_1  0.117617  7.980000e-01   \n",
       "109  ENSG00000167123      candidate_enh_3 -0.482057  8.800000e-02   \n",
       "\n",
       "            pair_type  \n",
       "0    positive_control  \n",
       "1    positive_control  \n",
       "2    positive_control  \n",
       "3    positive_control  \n",
       "4    positive_control  \n",
       "..                ...  \n",
       "105         discovery  \n",
       "106         discovery  \n",
       "107         discovery  \n",
       "108         discovery  \n",
       "109         discovery  \n",
       "\n",
       "[110 rows x 5 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(output_optional.uns[\"test_results\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03f0aab1",
   "metadata": {},
   "source": [
    "This is a data frame containing the same columns as the `pairs_to_test` data frame, plus at least one column containing a measure of the association for each pair. These columns can be `p_value`, `log2_fc`, `posterior_probability`, or any other measure of association.\n",
    "\n",
    "# Sample submission\n",
    "\n",
    "Here we present a sample Jamboree submission using PerTurbo to perform a Bayesian statistical analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9e6bdc50",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc72573d",
   "metadata": {},
   "source": [
    "\n",
    "## Function\n",
    "\n",
    "Here is a sample function that computes a posterior probability value based on PerTurbo\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "eef4f7be",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_scanpy_test_for_target(\n",
    "    mdata, intended_target_name, gene_ids=None, method=\"wilcoxon\"\n",
    "):\n",
    "    targeting_guides = (\n",
    "        mdata[\"guide\"].var[\"intended_target_name\"] == intended_target_name\n",
    "    )\n",
    "    targeted_cells = (\n",
    "        mdata[\"guide\"][:, targeting_guides].layers[\"guide_assignment\"].sum(axis=1)\n",
    "    )\n",
    "    if gene_ids is None:\n",
    "        adata = mdata[\"gene\"].copy()\n",
    "    else:\n",
    "        adata = mdata[\"gene\"][:, gene_ids].copy()\n",
    "    adata.obs = adata.obs.assign(\n",
    "        guide_status=[\"present\" if c > 0 else \"not_present\" for c in targeted_cells]\n",
    "    )\n",
    "    # sc.pp.log1p(mdata[\"gene\"])\n",
    "    # sc.pp.normalize_total(mdata[\"gene\"])\n",
    "    # sc.pp.regress_out(mdata[\"gene\"], \"prep_batch\")\n",
    "    sc.tl.rank_genes_groups(\n",
    "        adata,\n",
    "        groupby=\"guide_status\",\n",
    "        groups=[\"present\"],\n",
    "        method=method,\n",
    "        # inplace=False,\n",
    "    )\n",
    "    test_results_obj = adata.uns[\"rank_genes_groups\"]\n",
    "    test_results_df = pd.DataFrame(\n",
    "        {\n",
    "            \"gene_id\": test_results_obj[\"names\"][\"present\"],\n",
    "            \"intended_target_name\": intended_target_name,\n",
    "            \"p_value\": test_results_obj[\"pvals\"][\"present\"],\n",
    "            \"log2_fc\": test_results_obj[\"logfoldchanges\"][\"present\"],\n",
    "        }\n",
    "    )\n",
    "    return test_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "70068943",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>intended_target_name</th>\n",
       "      <th>p_value</th>\n",
       "      <th>log2_fc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000148303</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "      <td>3.059994e-03</td>\n",
       "      <td>5.771828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000114850</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "      <td>4.001935e-03</td>\n",
       "      <td>0.482913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000175854</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "      <td>5.329449e-03</td>\n",
       "      <td>0.473417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000095380</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "      <td>5.936656e-03</td>\n",
       "      <td>0.377006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000167112</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "      <td>7.017157e-03</td>\n",
       "      <td>0.337333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>ENSG00000148288</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "      <td>5.890089e-01</td>\n",
       "      <td>-0.074178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>ENSG00000228253</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "      <td>4.983945e-01</td>\n",
       "      <td>-0.181934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>ENSG00000106785</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "      <td>4.730550e-01</td>\n",
       "      <td>-0.209303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>ENSG00000165698</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "      <td>4.635750e-01</td>\n",
       "      <td>-0.542389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>ENSG00000147454</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "      <td>1.439739e-24</td>\n",
       "      <td>-3.986566</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_id intended_target_name       p_value   log2_fc\n",
       "0    ENSG00000148303      candidate_enh_1  3.059994e-03  5.771828\n",
       "1    ENSG00000114850      candidate_enh_1  4.001935e-03  0.482913\n",
       "2    ENSG00000175854      candidate_enh_1  5.329449e-03  0.473417\n",
       "3    ENSG00000095380      candidate_enh_1  5.936656e-03  0.377006\n",
       "4    ENSG00000167112      candidate_enh_1  7.017157e-03  0.337333\n",
       "..               ...                  ...           ...       ...\n",
       "107  ENSG00000148288      candidate_enh_1  5.890089e-01 -0.074178\n",
       "108  ENSG00000228253      candidate_enh_1  4.983945e-01 -0.181934\n",
       "109  ENSG00000106785      candidate_enh_1  4.730550e-01 -0.209303\n",
       "110  ENSG00000165698      candidate_enh_1  4.635750e-01 -0.542389\n",
       "111  ENSG00000147454      candidate_enh_1  1.439739e-24 -3.986566\n",
       "\n",
       "[112 rows x 4 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "run_scanpy_test_for_target(input_minimal, \"candidate_enh_1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "81e6f8b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    }
   ],
   "source": [
    "def run_scanpy_test_for_pairs(mdata, method=\"wilcoxon\"):\n",
    "    pairs_to_test = pd.DataFrame(mdata.uns[\"pairs_to_test\"])\n",
    "    targets = pairs_to_test[\"intended_target_name\"].unique()\n",
    "    test_results_list = []\n",
    "    for target in targets:\n",
    "        tested_genes = pairs_to_test.query(f\"intended_target_name=='{target}'\")\n",
    "        test_results_target = run_scanpy_test_for_target(\n",
    "            mdata, target, tested_genes[\"gene_id\"], method=method\n",
    "        )\n",
    "        test_results_list.append(test_results_target)\n",
    "    mdata.uns[\"test_results\"] = pd.concat(test_results_list).merge(pairs_to_test, how=\"right\")\n",
    "    return mdata\n",
    "\n",
    "\n",
    "mdata_output = run_scanpy_test_for_pairs(input_optional)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "6fa253d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>intended_target_name</th>\n",
       "      <th>p_value</th>\n",
       "      <th>log2_fc</th>\n",
       "      <th>pair_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000114850</td>\n",
       "      <td>ENSG00000114850</td>\n",
       "      <td>4.611026e-125</td>\n",
       "      <td>-4.046391</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000187109</td>\n",
       "      <td>ENSG00000187109</td>\n",
       "      <td>8.623445e-72</td>\n",
       "      <td>-8.093996</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000163866</td>\n",
       "      <td>ENSG00000163866</td>\n",
       "      <td>1.686899e-64</td>\n",
       "      <td>-2.530240</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000181610</td>\n",
       "      <td>ENSG00000181610</td>\n",
       "      <td>3.738848e-63</td>\n",
       "      <td>-3.105063</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000113552</td>\n",
       "      <td>ENSG00000113552</td>\n",
       "      <td>1.520147e-55</td>\n",
       "      <td>-2.536467</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000134851</td>\n",
       "      <td>ENSG00000134851</td>\n",
       "      <td>6.193244e-47</td>\n",
       "      <td>-1.858057</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000012660</td>\n",
       "      <td>ENSG00000012660</td>\n",
       "      <td>6.921034e-34</td>\n",
       "      <td>-1.996503</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSG00000147454</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "      <td>1.439739e-24</td>\n",
       "      <td>-3.986566</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000143321</td>\n",
       "      <td>ENSG00000143321</td>\n",
       "      <td>4.435129e-11</td>\n",
       "      <td>-1.215734</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000165702</td>\n",
       "      <td>candidate_enh_4</td>\n",
       "      <td>4.671225e-10</td>\n",
       "      <td>-1.052339</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000135046</td>\n",
       "      <td>candidate_enh_5</td>\n",
       "      <td>2.011033e-09</td>\n",
       "      <td>-2.456676</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSG00000106992</td>\n",
       "      <td>candidate_enh_3</td>\n",
       "      <td>1.029352e-07</td>\n",
       "      <td>-0.809274</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000122644</td>\n",
       "      <td>ENSG00000122644</td>\n",
       "      <td>2.153165e-07</td>\n",
       "      <td>-1.636916</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSG00000136938</td>\n",
       "      <td>candidate_enh_2</td>\n",
       "      <td>4.308644e-07</td>\n",
       "      <td>-2.916751</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000136929</td>\n",
       "      <td>candidate_enh_2</td>\n",
       "      <td>7.920648e-07</td>\n",
       "      <td>-1.205261</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000072274</td>\n",
       "      <td>ENSG00000072274</td>\n",
       "      <td>1.754460e-04</td>\n",
       "      <td>-0.815210</td>\n",
       "      <td>positive_control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>ENSG00000119333</td>\n",
       "      <td>candidate_enh_3</td>\n",
       "      <td>4.446526e-02</td>\n",
       "      <td>0.275923</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>ENSG00000148303</td>\n",
       "      <td>candidate_enh_3</td>\n",
       "      <td>4.549781e-02</td>\n",
       "      <td>-3.926848</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>ENSG00000147439</td>\n",
       "      <td>candidate_enh_1</td>\n",
       "      <td>7.230299e-02</td>\n",
       "      <td>0.406351</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSG00000136856</td>\n",
       "      <td>candidate_enh_3</td>\n",
       "      <td>9.742701e-02</td>\n",
       "      <td>0.520528</td>\n",
       "      <td>discovery</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_id intended_target_name        p_value   log2_fc  \\\n",
       "1   ENSG00000114850      ENSG00000114850  4.611026e-125 -4.046391   \n",
       "0   ENSG00000187109      ENSG00000187109   8.623445e-72 -8.093996   \n",
       "3   ENSG00000163866      ENSG00000163866   1.686899e-64 -2.530240   \n",
       "4   ENSG00000181610      ENSG00000181610   3.738848e-63 -3.105063   \n",
       "5   ENSG00000113552      ENSG00000113552   1.520147e-55 -2.536467   \n",
       "2   ENSG00000134851      ENSG00000134851   6.193244e-47 -1.858057   \n",
       "6   ENSG00000012660      ENSG00000012660   6.921034e-34 -1.996503   \n",
       "15  ENSG00000147454      candidate_enh_1   1.439739e-24 -3.986566   \n",
       "7   ENSG00000143321      ENSG00000143321   4.435129e-11 -1.215734   \n",
       "12  ENSG00000165702      candidate_enh_4   4.671225e-10 -1.052339   \n",
       "11  ENSG00000135046      candidate_enh_5   2.011033e-09 -2.456676   \n",
       "10  ENSG00000106992      candidate_enh_3   1.029352e-07 -0.809274   \n",
       "8   ENSG00000122644      ENSG00000122644   2.153165e-07 -1.636916   \n",
       "14  ENSG00000136938      candidate_enh_2   4.308644e-07 -2.916751   \n",
       "13  ENSG00000136929      candidate_enh_2   7.920648e-07 -1.205261   \n",
       "9   ENSG00000072274      ENSG00000072274   1.754460e-04 -0.815210   \n",
       "75  ENSG00000119333      candidate_enh_3   4.446526e-02  0.275923   \n",
       "93  ENSG00000148303      candidate_enh_3   4.549781e-02 -3.926848   \n",
       "44  ENSG00000147439      candidate_enh_1   7.230299e-02  0.406351   \n",
       "17  ENSG00000136856      candidate_enh_3   9.742701e-02  0.520528   \n",
       "\n",
       "           pair_type  \n",
       "1   positive_control  \n",
       "0   positive_control  \n",
       "3   positive_control  \n",
       "4   positive_control  \n",
       "5   positive_control  \n",
       "2   positive_control  \n",
       "6   positive_control  \n",
       "15         discovery  \n",
       "7   positive_control  \n",
       "12         discovery  \n",
       "11         discovery  \n",
       "10         discovery  \n",
       "8   positive_control  \n",
       "14         discovery  \n",
       "13         discovery  \n",
       "9   positive_control  \n",
       "75         discovery  \n",
       "93         discovery  \n",
       "44         discovery  \n",
       "17         discovery  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata_output.uns[\"test_results\"].sort_values(\"p_value\").head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "07481112",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_scanpy_test(input_mudata_fp, output_mudata_fp, method=\"wilcoxon\"):\n",
    "    mdata = md.read(input_mudata_fp)\n",
    "    mdata_out = run_scanpy_test_for_pairs(mdata, method=method)\n",
    "    mdata_out.write(output_mudata_fp)\n",
    "    return mdata_out"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4c5c725",
   "metadata": {},
   "source": [
    "\n",
    "## Demonstration\n",
    "\n",
    "Here is a demonstration of PerTurbo applied the Gasperini data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f0266ff3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:1301: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df[key] = c\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:1301: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df[key] = c\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:1301: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df[key] = c\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:1301: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df[key] = c\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:1301: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df[key] = c\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:1301: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df[key] = c\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n",
      "WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:1301: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df[key] = c\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:1301: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df[key] = c\n",
      "/Users/ljb80/.virtualenvs/perturbo-dev/lib/python3.11/site-packages/anndata/_core/anndata.py:1301: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df[key] = c\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'t-test': MuData object with n_obs × n_vars = 9704 × 167\n",
       "   obs:\t'prep_batch', 'within_batch_chip', 'within_chip_lane'\n",
       "   uns:\t'pairs_to_test', 'test_results'\n",
       "   2 modalities\n",
       "     gene:\t9704 x 112\n",
       "       obs:\t'num_expressed_genes', 'total_gene_umis'\n",
       "       var:\t'symbol', 'gene_chr', 'gene_start', 'gene_end'\n",
       "     guide:\t9704 x 55\n",
       "       obs:\t'num_expressed_guides', 'total_guide_umis'\n",
       "       var:\t'targeting', 'intended_target_name', 'intended_target_chr', 'intended_target_start', 'intended_target_end'\n",
       "       uns:\t'capture_method', 'moi'\n",
       "       layers:\t'guide_assignment',\n",
       " 'wilcoxon': MuData object with n_obs × n_vars = 9704 × 167\n",
       "   obs:\t'prep_batch', 'within_batch_chip', 'within_chip_lane'\n",
       "   uns:\t'pairs_to_test', 'test_results'\n",
       "   2 modalities\n",
       "     gene:\t9704 x 112\n",
       "       obs:\t'num_expressed_genes', 'total_gene_umis'\n",
       "       var:\t'symbol', 'gene_chr', 'gene_start', 'gene_end'\n",
       "     guide:\t9704 x 55\n",
       "       obs:\t'num_expressed_guides', 'total_guide_umis'\n",
       "       var:\t'targeting', 'intended_target_name', 'intended_target_chr', 'intended_target_start', 'intended_target_end'\n",
       "       uns:\t'capture_method', 'moi'\n",
       "       layers:\t'guide_assignment',\n",
       " 't-test_overestim_var': MuData object with n_obs × n_vars = 9704 × 167\n",
       "   obs:\t'prep_batch', 'within_batch_chip', 'within_chip_lane'\n",
       "   uns:\t'pairs_to_test', 'test_results'\n",
       "   2 modalities\n",
       "     gene:\t9704 x 112\n",
       "       obs:\t'num_expressed_genes', 'total_gene_umis'\n",
       "       var:\t'symbol', 'gene_chr', 'gene_start', 'gene_end'\n",
       "     guide:\t9704 x 55\n",
       "       obs:\t'num_expressed_guides', 'total_guide_umis'\n",
       "       var:\t'targeting', 'intended_target_name', 'intended_target_chr', 'intended_target_start', 'intended_target_end'\n",
       "       uns:\t'capture_method', 'moi'\n",
       "       layers:\t'guide_assignment'}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# break\n",
    "mudata_input_fp = f\"{data_dir}/gasperini_inference_input.h5mu\"\n",
    "\n",
    "output_scanpy = dict()\n",
    "for method in [\"t-test\", \"wilcoxon\", \"t-test_overestim_var\"]:\n",
    "    mudata_output_fp = f\"gasperini_inference_output_scanpy_{method}.h5mu\"\n",
    "    output_scanpy[method] = run_scanpy_test(mudata_input_fp, mudata_output_fp, method=method)\n",
    "output_scanpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e7f9fae1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "t-test\n",
      "     index          gene_id intended_target_name        p_value   log2_fc  \\\n",
      "1        1  ENSG00000114850      ENSG00000114850  1.680092e-121 -4.046391   \n",
      "5        5  ENSG00000113552      ENSG00000113552   4.495481e-71 -2.536467   \n",
      "3        3  ENSG00000163866      ENSG00000163866   1.913215e-67 -2.530240   \n",
      "0        0  ENSG00000187109      ENSG00000187109   1.268217e-60 -8.093996   \n",
      "4        4  ENSG00000181610      ENSG00000181610   4.334996e-51 -3.105063   \n",
      "2        2  ENSG00000134851      ENSG00000134851   6.910590e-48 -1.858057   \n",
      "6        6  ENSG00000012660      ENSG00000012660   4.269067e-38 -1.996503   \n",
      "109    109  ENSG00000147454      candidate_enh_1   1.019083e-27 -3.986566   \n",
      "68      68  ENSG00000135046      candidate_enh_5   4.361029e-19 -2.456676   \n",
      "85      85  ENSG00000165702      candidate_enh_4   8.165343e-19 -1.052339   \n",
      "7        7  ENSG00000143321      ENSG00000143321   6.059363e-12 -1.215734   \n",
      "66      66  ENSG00000106992      candidate_enh_3   1.718332e-09 -0.809274   \n",
      "100    100  ENSG00000136938      candidate_enh_2   2.074011e-08 -2.916751   \n",
      "99      99  ENSG00000136929      candidate_enh_2   2.663213e-08 -1.205261   \n",
      "8        8  ENSG00000122644      ENSG00000122644   2.838079e-08 -1.636916   \n",
      "9        9  ENSG00000072274      ENSG00000072274   9.981798e-05 -0.815210   \n",
      "98      98  ENSG00000136937      candidate_enh_2   1.041602e-02 -0.518276   \n",
      "101    101  ENSG00000147439      candidate_enh_1   1.113162e-02  0.406351   \n",
      "65      65  ENSG00000136854      candidate_enh_3   1.486150e-02 -0.506969   \n",
      "10      10  ENSG00000136856      candidate_enh_3   1.727239e-02  0.520528   \n",
      "\n",
      "            pair_type  \n",
      "1    positive_control  \n",
      "5    positive_control  \n",
      "3    positive_control  \n",
      "0    positive_control  \n",
      "4    positive_control  \n",
      "2    positive_control  \n",
      "6    positive_control  \n",
      "109         discovery  \n",
      "68          discovery  \n",
      "85          discovery  \n",
      "7    positive_control  \n",
      "66          discovery  \n",
      "100         discovery  \n",
      "99          discovery  \n",
      "8    positive_control  \n",
      "9    positive_control  \n",
      "98          discovery  \n",
      "101         discovery  \n",
      "65          discovery  \n",
      "10          discovery  \n",
      "wilcoxon\n",
      "     index          gene_id intended_target_name        p_value   log2_fc  \\\n",
      "1        1  ENSG00000114850      ENSG00000114850  4.611026e-125 -4.046391   \n",
      "0        0  ENSG00000187109      ENSG00000187109   8.623445e-72 -8.093996   \n",
      "3        3  ENSG00000163866      ENSG00000163866   1.686899e-64 -2.530240   \n",
      "4        4  ENSG00000181610      ENSG00000181610   3.738848e-63 -3.105063   \n",
      "5        5  ENSG00000113552      ENSG00000113552   1.520147e-55 -2.536467   \n",
      "2        2  ENSG00000134851      ENSG00000134851   6.193244e-47 -1.858057   \n",
      "6        6  ENSG00000012660      ENSG00000012660   6.921034e-34 -1.996503   \n",
      "109    109  ENSG00000147454      candidate_enh_1   1.439739e-24 -3.986566   \n",
      "7        7  ENSG00000143321      ENSG00000143321   4.435129e-11 -1.215734   \n",
      "85      85  ENSG00000165702      candidate_enh_4   4.671225e-10 -1.052339   \n",
      "68      68  ENSG00000135046      candidate_enh_5   2.011033e-09 -2.456676   \n",
      "66      66  ENSG00000106992      candidate_enh_3   1.029352e-07 -0.809274   \n",
      "8        8  ENSG00000122644      ENSG00000122644   2.153165e-07 -1.636916   \n",
      "100    100  ENSG00000136938      candidate_enh_2   4.308644e-07 -2.916751   \n",
      "99      99  ENSG00000136929      candidate_enh_2   7.920648e-07 -1.205261   \n",
      "9        9  ENSG00000072274      ENSG00000072274   1.754460e-04 -0.815210   \n",
      "10      10  ENSG00000119333      candidate_enh_3   4.446526e-02  0.275923   \n",
      "65      65  ENSG00000148303      candidate_enh_3   4.549781e-02 -3.926848   \n",
      "101    101  ENSG00000147439      candidate_enh_1   7.230299e-02  0.406351   \n",
      "11      11  ENSG00000136856      candidate_enh_3   9.742701e-02  0.520528   \n",
      "\n",
      "            pair_type  \n",
      "1    positive_control  \n",
      "0    positive_control  \n",
      "3    positive_control  \n",
      "4    positive_control  \n",
      "5    positive_control  \n",
      "2    positive_control  \n",
      "6    positive_control  \n",
      "109         discovery  \n",
      "7    positive_control  \n",
      "85          discovery  \n",
      "68          discovery  \n",
      "66          discovery  \n",
      "8    positive_control  \n",
      "100         discovery  \n",
      "99          discovery  \n",
      "9    positive_control  \n",
      "10          discovery  \n",
      "65          discovery  \n",
      "101         discovery  \n",
      "11          discovery  \n",
      "t-test_overestim_var\n",
      "     index          gene_id intended_target_name       p_value   log2_fc  \\\n",
      "1        1  ENSG00000114850      ENSG00000114850  3.261135e-52 -4.046391   \n",
      "0        0  ENSG00000187109      ENSG00000187109  2.082578e-31 -8.093996   \n",
      "3        3  ENSG00000163866      ENSG00000163866  3.839128e-28 -2.530240   \n",
      "5        5  ENSG00000113552      ENSG00000113552  5.401930e-26 -2.536467   \n",
      "4        4  ENSG00000181610      ENSG00000181610  1.096116e-24 -3.105063   \n",
      "2        2  ENSG00000134851      ENSG00000134851  1.277212e-23 -1.858057   \n",
      "6        6  ENSG00000012660      ENSG00000012660  6.170943e-17 -1.996503   \n",
      "109    109  ENSG00000147454      candidate_enh_1  1.179624e-10 -3.986566   \n",
      "85      85  ENSG00000165702      candidate_enh_4  3.681299e-08 -1.052339   \n",
      "7        7  ENSG00000143321      ENSG00000143321  3.239982e-06 -1.215734   \n",
      "66      66  ENSG00000106992      candidate_enh_3  5.172318e-05 -0.809274   \n",
      "68      68  ENSG00000135046      candidate_enh_5  1.061856e-04 -2.456676   \n",
      "100    100  ENSG00000136938      candidate_enh_2  1.716547e-04 -2.916751   \n",
      "8        8  ENSG00000122644      ENSG00000122644  1.944399e-04 -1.636916   \n",
      "99      99  ENSG00000136929      candidate_enh_2  3.218101e-04 -1.205261   \n",
      "9        9  ENSG00000072274      ENSG00000072274  8.194992e-03 -0.815210   \n",
      "101    101  ENSG00000147439      candidate_enh_1  5.026528e-02  0.406351   \n",
      "10      10  ENSG00000136856      candidate_enh_3  7.110161e-02  0.520528   \n",
      "98      98  ENSG00000136937      candidate_enh_2  9.679326e-02 -0.518276   \n",
      "65      65  ENSG00000136854      candidate_enh_3  1.065750e-01 -0.506969   \n",
      "\n",
      "            pair_type  \n",
      "1    positive_control  \n",
      "0    positive_control  \n",
      "3    positive_control  \n",
      "5    positive_control  \n",
      "4    positive_control  \n",
      "2    positive_control  \n",
      "6    positive_control  \n",
      "109         discovery  \n",
      "85          discovery  \n",
      "7    positive_control  \n",
      "66          discovery  \n",
      "68          discovery  \n",
      "100         discovery  \n",
      "8    positive_control  \n",
      "99          discovery  \n",
      "9    positive_control  \n",
      "101         discovery  \n",
      "10          discovery  \n",
      "98          discovery  \n",
      "65          discovery  \n"
     ]
    }
   ],
   "source": [
    "for method, mdata in output_scanpy.items():\n",
    "    print(method)\n",
    "    print(mdata.uns[\"test_results\"].sort_values(\"p_value\").head(20))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a1bbd10",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "warning,-all",
   "main_language": "python",
   "notebook_metadata_filter": "-all"
  },
  "kernelspec": {
   "display_name": "perturbo-dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
